server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -178,47 +185,80 @@ public void clear() throws InterruptedException
   }
 
   /**
-   * Add a row. Must not be called concurrently from multiple threads.
+   * Add a row.  This method may internally incur persisting data added so far.  Also, if too large data are persisted,
```

Comment:
I wonder if any problems could arise from introducing this blocking segment publishing step to add() (uploading segments could take a while).

If so, would it be feasible/better to do the segment publishing asynchronously (maybe complicated since you could still run into disk space issues unless the add/drain rates are managed effectively, but would allow the input to continue to be processed while the segment push to deep storage is occurring).